src/validate.ts

```diff
@@ -735,8 +735,43 @@ export function validate(
       return [
         newError(node, ErrorCodes.EventBus_Rules_do_not_support_Integrations),
       ];
+    } else if (
+      ts.isPropertyAccessExpression(node) ||
+      (ts.isIdentifier(node) && !ts.isTypeReferenceNode(node.parent))
+    ) {
+      let parent = node.parent;
+      while (
+        parent &&
+        (ts.isAsExpression(parent) || ts.isTypeAssertionExpression(parent))
+      ) {
+        // unroll `val as type` or `<type>val`
+        parent = parent.parent;
+      }
+      if (
+        // () => a
+        // () => a.b.c
+        ts.isFunctionDeclaration(parent) ||
+        ts.isFunctionExpression(parent) ||
+        ts.isArrowFunction(parent) ||
+        // a && a.b
+        // a || a.b
+        (ts.isBinaryExpression(parent) &&
+          (parent.operatorToken.kind ===
+            ts.SyntaxKind.AmpersandAmpersandToken ||
+            parent.operatorToken.kind === ts.SyntaxKind.BarBarToken)) ||
+        // !a
+        // !a.b
+        (ts.isPrefixUnaryExpression(parent) &&
+          parent.operator === ts.SyntaxKind.ExclamationToken)
```

Comment:
Wouldn't falsey work? `[{ exists: false }, 0, "", false]`

Reply:
Deferring this to https://github.com/functionless/functionless/issues/365